[REEF-1260] Adding a sample for Context Start handler

jwang98052 opened the pull request:
Added a sample of ContextStartHandler for Fault Tolerant as a test case

JIRA: [REEF-1260](https://issues.apache.org/jira/browse/REEF-1260)

This closes #

tcNickolas commented:
Travis CI fails because license is missing in the test file. Please add

markusweimer commented on lang/cs/Org.Apache.REEF.Tests/Functional/FaultTolerant/TestContextStart.cs:
```diff
@@ -0,0 +1,201 @@
+
+        public TestContextStart()
+        {
+            Init();
```
Isn't there a specific convention for test initialization code in xUnit?

jwang98052 replied:
This is the pattern every test follows so far. If there is anything, we can change it consistently.

tcNickolas replied:
The convention is to put test initialization in class constructor.
I have some ideas about cleaning up tests inherited from ReefFunctionalTest: all constructors call Init() so they can be replaced with base constructor calling Init, and same about Dispose(). But I want to postpone this change until our AppVeyor config is good, so that we have more validation of this approach.

markusweimer replied:
Hmm, that creates an abstract base class with all the associated problems. Let's take the discussion to that JIRA :)

tcNickolas replied:
https://issues.apache.org/jira/browse/REEF-1265

markusweimer commented on lang/cs/Org.Apache.REEF.Tests/Functional/FaultTolerant/TestContextStart.cs:
```diff
@@ -0,0 +1,201 @@
+                    Logger.Log(Level.Info, DataDownLoadCompletedMessage);
+                    return null;
+                }
+                Thread.Sleep(5 * 1000);
```
Why the need for the sleep?

jwang98052 replied:
We usually sleep little bit to simulate task doing something. But not necessary in this case.

jwang98052 commented:
@markusweimer i have pushed the change with the review comments addressed.

jwang98052 commented:
Hold on, I need to update a test case.

jwang98052 commented:
Ready for next review.

markusweimer commented:
I'll do another pass.

markusweimer commented on lang/cs/Org.Apache.REEF.Tests/Functional/FaultTolerant/TestContextStart.cs:
```diff
@@ -0,0 +1,199 @@
+                _requestor.Submit(_requestor.NewBuilder().Build());
+            }
+
+            public void OnNext(IActiveContext value)
```
You could validate that the context has the right id here.

jwang98052 replied:
Done

markusweimer commented on lang/cs/Org.Apache.REEF.Tests/Functional/FaultTolerant/TestContextStart.cs:
```diff
@@ -0,0 +1,199 @@
+        }
+
+        /// <summary>
+        /// This test case submit a context with a Context start handler and do something is the handler
```
`is` -> `in`

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Tests/Functional/FaultTolerant/TestContextStart.cs:
```diff
@@ -0,0 +1,199 @@
+            private readonly DataDownLoader _dataDownLoader;
+
+            [Inject]
+            private ContextStartHandler(DataDownLoader dataDownLoader)
```
Do we need to download data here? Wouldn't it be good enough to just note the fact that `OnNext` has been invoked?

jwang98052 replied:
It is to show we can inject any object to do something with the start handler. I renamed it to make it more generic.

markusweimer commented:
Done with a pass. A general question: We validate these tests based on the logs they write, right? This shouldn't stop us from merging this, but we really should re-evaluate that choice. Testing on YARN this way would be way too slow.

jwang98052 commented:
It is true a lot of validation is based on the messages in logs. If we can come out any better way, we can change cross the board.

jwang98052 commented:
@markusweimer 
I have addressed your comments.
I also enhanced the test scenario by adding one more layer of context, added task complete and context close handlers
More asserts are added to validate the events received.

markusweimer commented on lang/cs/Org.Apache.REEF.Tests/Functional/FaultTolerant/TestContextStart.cs:
```diff
@@ -0,0 +1,238 @@
+                value.ActiveContext.Dispose();
+            }
+
+            public void OnNext(IClosedContext value)
```
This shouldn't pass. The first context closed has id `ContextId2`, and this passes. However, the second closed context has id `ContextId1`. Hence, this code should fail.

markusweimer replied:
I tested this and the test passes. This is bad news, as it shouldn't pass. Or am I missing something?

markusweimer replied:
I tested this and the test passes. This is bad news, as it shouldn't pass. Or am I missing something?

jwang98052 replied:
@markusweimer We don't receive second IClosedContext event. I added ICompletedEvaluator handler, it shows the evaluator is completed. There may be a bug some where. I will open a Jira for it. 

The change has pushed in. 

I updated he code to handle the second close.

markusweimer replied:
The new code isn't on GitHub yet. Can you double-check that the push went through?

markusweimer replied:
I see it now. However, as I reported on REEF-1269, I was wrong: We shouldn't actually get an `IClosedContext` for the root context. The java code doesn't create the event. I think that is a mistake and will file another JIRA for that.

For now, we should revert this code to assume that there is no `IClosedContext` for the root context.

jwang98052 replied:
Done

jwang98052 commented:
Network issue. It is in now.

jwang98052 commented:
@markusweimer I have push the change to revert the close context code.

markusweimer commented:
LGTM. Will test and merge.

markusweimer commented:
Tests pass. will merge.

markusweimer commented:
I merged it. GitHub seems to take some time with the update here.
